File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.java
class Solution {

    int help(int idx, int[] prices , int[][][]dp, int buy, int cnt){
      if(idx == prices.length || cnt == 2) return 0;
    

      if(dp[idx][buy][cnt] != -1) return dp[idx][buy][cnt];

      if(buy == 1){
        return dp[idx][buy][cnt] = Math.max(-prices[idx] + help(idx+1,prices,dp,0,cnt), help(idx+1,prices,dp,1,cnt));

      }
     

      return dp[idx][buy][cnt] = Math.max(prices[idx] + help(idx+1,prices,dp,1,cnt+1), help(idx+1,prices,dp,0,cnt));
    }
    public int maxProfit(int[] prices) {
        int[][][]dp = new int[prices.length][2][2];
        
        for(int[][] a: dp){
          for(int[] b: a){
           Arrays.fill(b,-1);
        }
        }

        return help(0,prices,dp,1,0);
    }
}

File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.cpp
class Solution {
public:

    int helper(int idx ,int cnt,int buy,vector<int>& prices,vector<vector<vector<int>>>&dp ){

        if(idx == prices.size()){
            return 0;
        }
            if(dp[idx][buy][cnt] != -1) return dp[idx][buy][cnt];
        int profit =0;
        if(buy == 1 && cnt>0){
             profit =  max(-prices[idx] + helper(idx+1,cnt,0,prices,dp),helper(idx+1,cnt,1,prices,dp));
        }
        else if(cnt>0){
          profit = max(prices[idx] + helper(idx+1,cnt-1,1,prices,dp),helper(idx+1,cnt,0,prices,dp));
        }
        return dp[idx][buy][cnt] = profit;
    }
    int maxProfit(vector<int>& prices) {
        vector<vector<vector<int>>>dp(prices.size(),vector<vector<int>>(2,vector<int>(3,-1)));
      
        return helper(0,2,1,prices,dp);
    }
};